composer/algorithms/gradient_clipping/gradient_clipping.py

```diff
@@ -0,0 +1,230 @@
+            gc = GradientClipping(clipping_type='norm', clipping_threshold=0.1)
+            trainer = Trainer(
+                model=model,
+                train_dataloader=train_dataloader,
+                eval_dataloader=eval_dataloader,
+                max_duration="1ep",
+                algorithms=[gc],
+                optimizers=[optimizer]
+            )
+    Args:
+        parameters (torch.Tensor or Iterable[torch.Tensor]): The parameters to of the
+            model for whose gradients we will clip
+        clipping_type ('adaptive', 'norm', 'value'): String denoting which type of
+            gradient clipping to do. The options are:
+
+            =============   ===========                             ========
+            clipping_type   description                             function
+            =============   ===========                             ========
+            'adaptive'      Clips all gradients based on            composer.algorithms.
+                             gradient norm:parameter norm            gradient_clipping.
+                             ratio.                                  gradient_clipping.
+                                                                    _apply_agc
+
+            'norm'          Clips gradient norm                     torch.nn.utils.
+                                                                    clip_grad_norm_
+
+            'value'         Clips gradient at specified value       torch.nn.utils.
+                                                                    clip_grad_value_
+            =============
+
+        clipping_threshold (float, optional): Specifies:
+            * what value to clip the gradients to (for 'value')
+            * what values to clip the gradient norms to (for 'norm')
+            * threshold by which if grad_norm / weight_norm is greater than this threshold
+                then scale gradients by
+                this threshold * (weight_norm / grad_norm) (for 'adaptive')
+    
+    Raises:
+        NotImplementedError: if deepspeed is enabled and clipping_type is not 'norm'.
+    """
+
+    def __init__(self, clipping_type: str, clipping_threshold: float):
+        self.clipping_type = clipping_type
+        self.clipping_threshold = clipping_threshold
+        try:
+            import deepspeed
+            self.deep_speed_imported = True
+        except ImportError:
+            self.deep_speed_imported = False
+
+    def match(self, event: Event, state: State) -> bool:
+        """Run on ``Event.AFTER_TRAIN_BATCH``."""
+        return event == Event.AFTER_TRAIN_BATCH
+
+    def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
+        # If Deepspeed is enabled and clipping_type is norm, set the clipping through
+        # deepspeed.
+        if state.is_model_deepspeed:
+            if self.clipping_type == 'norm':
+                state.model._config.gradient_clipping = self.clipping_threshold
```

Comment:
I don't have the final say here, but I'm actually somewhat in favor of leaving it as currently written. Having special code for an algorithm in the trainer feels really wrong, and I don't want there to be a precedent for that.

On the other hand, DeepSpeed is such a mess that I doubt this will be the last time we access its private variables in this way (getting it to work with adaptive grad accum is a place where we might do something similar).

Reply:
fwiw, agreed with @jbloxham 's take here

Reply:
yeah seems like a lot just to avoid a private var

Reply:
going to keep it as is

Reply:
I'm really not a fan of this, as it could cause a breaking change without any errors when we upgrade the deepspeed version. In addition, how do we know this works? Deepspeed could be reading this variable during `deepspeed.initialize`, in which case this monekypatch would have no effect.

One alternative would be to put the `deepspeed_config` on the state during `Event.INIT`, so algorithms can modify it directly.

If the decision has already been made to modify this private variable, please add a test case to ensure that the gradients are indeed being clipped (not just a functional test, but a test to ensure that doing it this way actually clips the gradients).

Reply:
> I'm really not a fan of this, as it could cause a breaking change without any errors when we upgrade the deepspeed version. 
I'm not really convinced by this argument since plenty of our existing code could break if we upgrade deepspeed anyway. Not sure how this would be any different.

>In addition, how do we know this works? Deepspeed could be reading this variable during `deepspeed.initialize`, in which case this monekypatch would have no effect.

Good call. It should actually modify the optimizer directly. Fixed here:  3b66190

> One alternative would be to put the `deepspeed_config` on the state during `Event.INIT`, so algorithms can modify it directly. 

That is a certainly a valid solution, but would be more work and seemingly not necessary.

> If the decision has already been made to modify this private variable, please add a test case to ensure that the gradients are indeed being clipped (not just a functional test, but a test to ensure that doing it this way actually clips the gradients).

I think just verifying that the optimizer has optimizer.clip_grad == clipping_threshold is enough since we can trust that deepspeed works as written and that their tests are good.

Reply:
ok that test is here: 
fdbb567